services/job_subscriber.go

```diff
@@ -71,14 +95,69 @@ func (js *JobSubscriber) Disconnect() {
 }
 
 // OnNewHead resumes all pending job runs based on the new head activity.
-func (js *JobSubscriber) OnNewHead(head *models.BlockHeader) {
-	pendingRuns, err := js.Store.JobRunsWithStatus(models.RunStatusPendingConfirmations)
+func (js *jobSubscriber) OnNewHead(head *models.BlockHeader) {
+	pendingRuns, err := js.Store.JobRunsWithStatus(models.RunStatusPendingConfirmations, models.RunStatusInProgress)
 	if err != nil {
 		logger.Error(err.Error())
 	}
+
+	activeJobRunIDs := make(map[string]struct{})
+
+	js.workerMutex.Lock()
+	defer js.workerMutex.Unlock()
 	for _, jr := range pendingRuns {
-		if _, err := ExecuteRunAtBlock(jr, js.Store, jr.Result, head.ToIndexableBlockNumber()); err != nil {
-			logger.Error(err.Error())
+		activeJobRunIDs[jr.ID] = struct{}{}
+
+		workerChannel := js.WorkerChannelFor(jr)
+		blockNumber := head.ToIndexableBlockNumber()
+		workerChannel <- blockNumber
+	}
+
+	//Stop any workers that didn't have corresponding pending confirmations
+	for id, workerChannel := range js.workers {
+		if _, ok := activeJobRunIDs[id]; !ok {
+			close(workerChannel)
+			delete(js.workers, id)
 		}
 	}
 }
+
+// workerChannelFor accepts a JobRun and returns a worker channel dedicated
+// to that JobRun. The channel accepts new block heights for triggering runs,
+// and ensures that the block height confirmations are run syncronously.
+func (js *jobSubscriber) WorkerChannelFor(jr models.JobRun) chan *models.IndexableBlockNumber {
+	workerChannel, present := js.workers[jr.ID]
+	if !present {
+		workerChannel = make(chan *models.IndexableBlockNumber, 100)
+		js.workers[jr.ID] = workerChannel
+
+		go func() {
+			js.workerWaiter.Add(1)
+			defer js.workerWaiter.Done()
```

Comment:
I think my vision was to combine this with the `executeRunQueue` in `RunManager` above. But I don't know the nuances of the package imports causing cycles and the available fields to the callers. I understand why `RunQueue` is in `store` while `RunManager` is in application, but find it fascinating.

Reply:
Managing which runs are pending and which runs aren't seemed like logic that might be better fit for the job subscriber than the run manager. That, the circular dependency problem, and the fact that this is one of the few places where we use a block height when executing a run made it feel like a bit of a forced abstraction. I think it might be helpful to try and coerce a couple of the other seven scenarios before tying this one.

Reply:
I agree managing pending runs should belong in `JobSubscriber`. But isn't that handled by `OnNewHead` above? That could feed into the `RunQueue`. When I said "combine this" I was speaking more specifically about the workerChannel (channel specific to a job run). It does seem like the BlockNumber is holding you up but that could be a second parameter or adding to `input RunResult`. Food for thought.